### PR TITLE
Fix vm tests

### PR DESCRIPTION
**Pull request by petrkotas**
**What this PR does / why we need it**:
Rework the VM lifecycle tests. It fixes the naming in the tests to be consistent and differentiate
between VM and VMI. The gomega matchers are now decorated with additional descriptions
making the debugging failed tests easier. Finally the eventually loops are now less time demanding
and in few cases check for equivalent events.

Overall this PR makes vm-tests faster and more stable.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
One of the patches fixing the overall CI.

**Release note**:

```release-note
NONE
```

---
This change is [Reviewable](https://reviewable.io/reviews/kubevirt/kubevirt/1523)

### CODE REVIEW

**Comment by petrkotas** [comment]
@vladikr @stu-gott can you please take a look at this one? I would like to merge as soon as possible so I can continue with writing the guide for fixing the rest. Thanks

File: tests/vm_test.go [hunk 1]
```diff
@@ -269,38 +270,38 @@ var _ = Describe("VirtualMachine", func() {
...
+				firstVMI, err = virtClient.VirtualMachineInstance(newVM.Namespace).Get(newVM.Name, &v12.GetOptions{})
 				if err != nil {
 					return err
 				}
 				if !firstVMI.IsRunning() {
 					return fmt.Errorf("vmi still isn't running")
 				}
 				return nil
-			}, 120*time.Second, 1*time.Second).Should(Succeed())
+			}, 120*time.Second, 5*time.Second).Should(Succeed(), "VMI should have started")
```
Comment (rmohr) [inline tests/vm_test.go]: Did you observe specific issues here? Is one call per second really too much?

**Comment by rmohr** [comment]
@petrkotas that one does not need a release note.

File: tests/vm_test.go (hunk 1 above)
Comment (petrkotas) [reply tests/vm_test.go]: I have two reasons for this setup. First there is simply no need to check it faster. The VM takes time to boot. Second, I am thinking of parallel tests. I do not want to DOS API when running in parallel.

Do you agree?

**Comment by rmohr** [comment]
for reference: https://jenkins.ovirt.org/job/kubevirt_kubevirt_standard-check-pr/1962/artifact/ci_build_summary.html

**Comment by rmohr** [comment]
ci test please

**Comment by rmohr** [comment]
@petrkotas I see pretty often 

```
[Fail] VirtualMachine A valid VirtualMachine given [It] should remove VirtualMachineInstance once the VM is marked for deletion 
/root/go/src/kubevirt.io/kubevirt/tests/vm_test.go:214
```

fail now. Can it be that the timeout changes cause this?

**Comment by petrkotas** [comment]
@rmohr investigating the issue.

**Comment by petrkotas** [comment]
@rmohr I fixed it. I made the test check for delete timestamp, but forgot the vmi could be already gone. Fixed.

File: tests/vm_test.go [hunk 2]
```diff
@@ -119,147 +119,152 @@ var _ = Describe("VirtualMachine", func() {
...
 			Eventually(func() bool {
-				_, err := virtClient.VirtualMachine(newVMI.Namespace).Get(newVMI.Name, &v12.GetOptions{})
-				if errors.IsNotFound(err) {
-					return true
-				}
-				return false
-			}, 300*time.Second, 1*time.Second).Should(BeTrue())
+				vm, err := virtClient.VirtualMachine(newVM.Namespace).Get(newVM.Name, &v12.GetOptions{})
+				Expect(err).ToNot(HaveOccurred(), "Should get the VM")
+				return vm.GetObjectMeta().GetDeletionTimestamp() != nil
```
Comment (davidvossel) [inline tests/vm_test.go]: nope, check for ```errors.IsNotFound(err)``` instead

File: tests/vm_test.go [hunk 3]
```diff
@@ -119,147 +119,152 @@ var _ = Describe("VirtualMachine", func() {
...
+			}, 30*time.Second, 2*time.Second).ShouldNot(BeEmpty(), "Should have the ownerReference")
 
 			// Delete it
 			orphanPolicy := v12.DeletePropagationOrphan
-			Expect(virtClient.VirtualMachine(newVMI.Namespace).
-				Delete(newVMI.Name, &v12.DeleteOptions{PropagationPolicy: &orphanPolicy})).To(Succeed())
-			// Wait until the virtual machine is deleted
+			Expect(virtClient.VirtualMachine(newVM.Namespace).
+				Delete(newVM.Name, &v12.DeleteOptions{PropagationPolicy: &orphanPolicy})).To(Succeed(), "Should be able to delete VM")
+			// Wait until the offlinevmi is deleted
```
Comment (davidvossel) [inline tests/vm_test.go]: s/offlinevmi/vm

File: tests/vm_test.go [hunk 4]
```diff
@@ -119,147 +119,152 @@ var _ = Describe("VirtualMachine", func() {
...
-				_, err := virtClient.VirtualMachine(newVMI.Namespace).Get(newVMI.Name, &v12.GetOptions{})
-				if errors.IsNotFound(err) {
-					return true
-				}
-				return false
-			}, 300*time.Second, 1*time.Second).Should(BeTrue())
+				vm, err := virtClient.VirtualMachine(newVM.Namespace).Get(newVM.Name, &v12.GetOptions{})
+				Expect(err).ToNot(HaveOccurred(), "Should get the VM")
+				return vm.GetObjectMeta().GetDeletionTimestamp() != nil
+			}, 30*time.Second, 2*time.Second).Should(BeTrue(), "VM should disappear")
```
Comment (davidvossel) [inline tests/vm_test.go]: might take longer than 30 seconds

File: tests/vm_test.go [hunk 5]
```diff
@@ -436,69 +441,68 @@ var _ = Describe("VirtualMachine", func() {
...
-				By("Ensuring VMI is running")
-				Eventually(func() bool {
-					newVMI, err = virtClient.VirtualMachine(newVMI.Namespace).Get(newVMI.Name, &v12.GetOptions{})
-					Expect(err).ToNot(HaveOccurred())
-					return newVMI.Status.Ready
-				}, 360*time.Second, 1*time.Second).Should(BeTrue())
+				By("Ensuring VM is running")
+				Eventually(func() error {
+					_, err = virtClient.VirtualMachine(newVM.Namespace).Get(newVM.Name, &v12.GetOptions{})
+					return err
```
Comment (davidvossel) [inline tests/vm_test.go]: This test is no longer accurate with these changes.

checking for a GET error does not occur is not the same as ensuring newVMI.Status.Ready == true.  This is just checking that the VM exists, not that it's running. The previous check was right

**Comment by rmohr** [comment]
@davidvossel good review!

**Comment by petrkotas** [comment]
@davidvossel thanks for clarifications. Fixing it. Thinking about this, I would like to see information explaining what the tests verifies directly in the test itself. The current description is not enough.
Will figure out how to fix that.

**Comment by petrkotas** [comment]
Closing this, as I can no longer see the failures

**Pull request closed without merging**